opt_adapt/opt.py

```diff
@@ -52,11 +53,33 @@ def __init__(self, options={}):
                 raise ValueError(f"Option {key} not recognised")
             self.__setattr__(key, value)
 
+# Only works for problem with control paramter in r-space
+def line_search(forward_run, mesh, u, P, J, dJ, Rspace, alpha=1e-4, max_search_iter=100):
+    
+    lr = 1
+    if Rspace:
+        initial_slope = float(dJ) * float(P)
+    else:
+        initial_slope = np.dot(dJ.dat.data, P.dat.data)
+
+    if initial_slope==0.0:
+        return 1.0
+    
+    for i in range(max_search_iter):
+        u_plus = u + lr*P 
+        J_plus, u_plus = forward_run(mesh, u_plus)
+        # check Armijo rule:
+        if J_plus-J <= alpha*lr*initial_slope:
+            break
+        lr /= 2
```

Comment:
It would be good to put `lr *= tau`, where `tau` is another keyword argument. You could still make it default to 0.5, but it gives us more flexibility to try and figure out the best option.